src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java

```diff
@@ -0,0 +1,203 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.BiConsumer;
+
+/**
+ * @author Georg Held
+ */
+public class EmbeddedHiveMQImpl implements EmbeddedHiveMQ {
+
+    private static final @NotNull Logger log = LoggerFactory.getLogger(EmbeddedHiveMQImpl.class);
+
+    private final @NotNull SystemInformationImpl systemInformation;
+    private final @NotNull MetricRegistry metricRegistry;
+    private @Nullable FullConfigurationService configurationService;
+    private @Nullable Injector injector;
+
+    private volatile @NotNull State currentState;
+    private volatile @Nullable CompletableFuture<Void> startFuture;
+    private volatile @Nullable CompletableFuture<Void> stopFuture;
+
+    EmbeddedHiveMQImpl(
+            final @Nullable File conf, final @Nullable File data, final @Nullable File extensions) {
+
+        currentState = State.NOT_STARTED;
+        systemInformation = new SystemInformationImpl(true, true, conf, data, extensions);
+        // we create the metric registry here to make it accessible before start
+        metricRegistry = new MetricRegistry();
+    }
+
+    private enum State {
+
+        NOT_STARTED,
+        STARTING,
+        RUNNING,
+        STOPPING,
+        STOPPED;
+
+        private boolean isHigher(final @NotNull State other) {
+            return this.ordinal() > other.ordinal();
+        }
+
+        private boolean isLower(final @NotNull State other) {
+            return this.ordinal() < other.ordinal();
+        }
+    }
+
+    private synchronized void advanceState(final @NotNull State newState) {
+        if ((currentState == State.STOPPED && newState == State.NOT_STARTED)
+                || newState.isHigher(currentState)) {
+            log.debug("Advancing EmbeddedHiveMQState from \"{}\" to \"{}\"", currentState, newState);
+            currentState = newState;
+        }
+    }
+
+    @Override
+    public synchronized @NotNull CompletableFuture<Void> start() {
+        advanceState(State.NOT_STARTED);
+        if (currentState == State.NOT_STARTED) {
+            advanceState(State.STARTING);
+            log.info("Starting EmbeddedHiveMQ.");
+
+            startFuture = new CompletableFuture<>();
+            CompletableFuture.runAsync(() -> {
+
+                configurationService = ConfigurationBootstrap.bootstrapConfig(systemInformation);
+
+                bootstrapInjector();
+                final HiveMQServer hiveMQServer = injector.getInstance(HiveMQServer.class);
+                try {
+                    hiveMQServer.start();
+                } catch (final Exception e) {
+                    throw new HiveMQServerException(e);
+                }
+                advanceState(State.RUNNING);
+            }).whenComplete(HiveMQServerException.handler(startFuture));
+        } else {
+            log.warn("Could not start EmbeddedHiveMQ. Reason: \"Already started\"");
```

Comment:
imho this is weird, if we call start after stop we will only get a log, but the old startFuture which will return success.
I think this case should return an errored future.
If we want to expose the startFuture, we should add an extra method with clear naming (like startedFuture) and clear documentation. "start" includes the actual action to start HiveMQ.

Reply:
hm, I agree. The stop now works while the start is still in progress but not vice versa. Would it be good if both could be triggered while the other is in progress?
Like the *level based* system of kubernetes?

Reply:
I implemented this.